planck-c/functions.c

```diff
@@ -1030,6 +1030,43 @@ JSValueRef function_list_files(JSContextRef ctx, JSObjectRef function, JSObjectR
     return JSValueMakeNull(ctx);
 }
 
+JSValueRef function_mktemp(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject,
+                           size_t argc, const JSValueRef args[], JSValueRef *exception) {
+    if (argc == 1
+        && JSValueGetType(ctx, args[0]) == kJSTypeBoolean) {
+
+        bool directory = JSValueToBoolean(ctx, args[0]);
+
+        char* tmpdir = getenv("TMPDIR");
+        if (!tmpdir) {
+            tmpdir = "/tmp";
+        }
+
+        char template[PATH_MAX];
+        if (str_has_suffix(tmpdir, "/") == 0) {
+            snprintf(template, PATH_MAX, "%splanck.XXXXXX", tmpdir);
+        } else {
+            snprintf(template, PATH_MAX, "%s/planck.XXXXXX", tmpdir);
+        }
+
+        char *temp_name;
+        if (directory) {
+            temp_name = mkdtemp(template);
+        } else {
+            temp_name = mktemp(template);
```

Comment:
GCC is complaining about the use of `mktemp()` here (apparently because of [the security risk](https://linux.die.net/man/3/mktemp)). The alternative recommended is `mkstemp()`, which creates the temporary file. If we go down this route, the directory branch that uses `mkdtemp()` will need to be updated to create the directory (and `planck.io/temp-file` and `planck.io/temp-directory` will need to be updated accordingly as `as-file` will no longer be required).

Reply:
@pyrmont The additional patch below is a different stab at this. If we were to use `mkstemp()`, this would result in a file descriptor being returned. While this descriptor could be used in the I/O subsystem to, say set up a reader, writer, input-stream, or output-stream, things are really set up to start off with a file (name). Another issue with the descriptor approach is that you'd probably need to first set up, say, an output-stream, write to it, and then somehow reset the file descriptor and set up an input stream on it to read back from it.

The alternative (but arguably less than ideal) approach taken in the other subsequent patch is to still attempt to use `mktemp()`, but to spin loop in order to mimic what `mkstemp()` would do, while still returning a file name.

This results in something that feels a lot like Java's `createTempFile`: https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createTempFile(java.lang.String,%20java.lang.String,%20java.nio.file.attribute.FileAttribute...)

Reply:
A compiler warning isn't the end of the world but the revised code still causes one to be raised and so I wondered if there were a way to use `mkstemp()` but still be oriented around a filename rather than a file descriptor.

My reading of the docs is that `mkstemp()` (and `mkdtemp()` for that matter) mutate the character array used as the template (this is the reason the template has to be expressed as a character array rather than as a string literal). We can take advantage of this to retain our filename orientation even if a filename isn't expressly returned by the function.

Here's an alternative:

```clj
JSValueRef function_mktemp(JSContextRef ctx, JSObjectRef function, JSObjectRef thisObject,
                           size_t argc, const JSValueRef args[], JSValueRef *exception) {
    if (argc == 1
        && JSValueGetType(ctx, args[0]) == kJSTypeBoolean) {

        bool directory = JSValueToBoolean(ctx, args[0]);

        char* tmpdir = getenv("TMPDIR");
        if (!tmpdir) {
            tmpdir = "/tmp";
        }

        char template[PATH_MAX];
        if (str_has_suffix(tmpdir, "/") == 0) {
            snprintf(template, PATH_MAX, "%splanck.XXXXXX", tmpdir);
        } else {
            snprintf(template, PATH_MAX, "%s/planck.XXXXXX", tmpdir);
        }

        char *temp_name = template;

        bool error;
        if (directory) {
            error = (NULL == mkdtemp(template));
        } else {
            error = (-1 == mkstemp(template));
        }

        if (!error) {
            return c_string_to_value(ctx, temp_name);
        } else {
            JSValueRef arguments[1];
            arguments[0] = c_string_to_value(ctx, strerror(errno));
            *exception = JSObjectMakeError(ctx, 1, arguments, NULL);
        }

    }
    return JSValueMakeNull(ctx);
}
```

What do you think? It seemed to test fine and I thought had the advantage of being simpler.

I confess I didn't fully understand what you meant when you said:

> Another issue with the descriptor approach is that you'd probably need to first set up, say, an output-stream, write to it, and then somehow reset the file descriptor and set up an input stream on it to read back from it.

Perhaps the above code still has this problem.

Reply:
@pyrmont Excellent! I had noticed that as well, but it didn't click for me that this was a way to access the filename. I'll revise the patch reflecting that approach. (I think the file descriptor needs to be closed; you'll see that in the revision I plan on making.)

Reply:
@mfikes Ahhhh... right, of course. The descriptor being left open could cause issues. That makes sense now!

Reply:
@pyrmont Yeah, and the bit above you said you didn't fully understand: It has to do with the question of what do you do if you try to make this function actually return the file descriptor itself (as opposed to the file name). That alternative leads to the difficulty I was speaking to, and thus the desire to stick with returning the file name somehow.

Reply:
```c
if (temp_name) {
    return c_string_to_value(ctx, temp_name);
} else {
    JSValueRef arguments[1];
    arguments[0] = c_string_to_value(ctx, strerror(errno));
    *exception = JSObjectMakeError(ctx, 1, arguments, NULL);
}
```

The new code looks pretty good to me but I'm not sure testing the value of `temp_name` will necessarily produce the right result here. `mkdtemp()` sets the value of the character array to `NULL` if there's an error but that [doesn't look](https://linux.die.net/man/3/mkstemp) to be the behaviour of `mkstemp()`. That was the reason I was using that `error` boolean in the alternative.

Reply:
I have the code initializing `temp_name` to `NULL` now, and on the branch working with `mkstemp()` it only assigns to `temp_name` if `mkstemp()` returns a usable file descriptor that can also be successfully closed.

So, in short I think the logic in the latest commit is correct. But let me know if you see a path where `temp_name` can be incorrect.

Reply:
D'oh—you're right. My bad!

Reply:
Just added one more commit "clearer assignment to temp_name" that simplifies reasoning about the logic.